Improved validation layer

kirrg001 opened the pull request:
no issue

Please take a look at the commits. Everything is described.

Summary of the major problems:
You won't receive a validation error if you add a user without an email. This is caused by a bug in our validation layer. Described in the target commit.

If you add a user without name and without a slug, the slug generation does not work and breaks. This must be prevented by requiring a name.

**This PR contains unit tests 🙊**
I did a quick browser test, looks good.

kirrg001 committed:
Validate schema based on the bookshelf model

no issue

- the `validateSchema` helper was broken
  - if you add a user without email, you will receive a database error
  - but the validation error should catch that email is passed with null
- it was broken, because:
  - A: it called `toJSON` -> this can remove properties from the output (e.g. password)
  - B: we only validated fields, which were part of the JSON data (model.hasOwnProperty)
- fixed on broken test
  - if you import a post without a status, it should not error
  - it falls back to the default value

kirrg001 committed:
Added unit test for models/user

no issue

- proof that validations work now as expected

kirrg001 committed:
Tidied up model defaults

no issue

- the database and knex allows us to use `defaultTo`
  - we make use of it
  - this value is stored in the database
- the bookshelf `defaults` function has the purpose to generate dynamic values
- the static values are already defined on the database level
- redundant data

kirrg001 committed:
Make use of `onValidate` from the base model

no issue

- always call the base class, except you have a good reason
- no need to reimplement logic

ErisDS commented on core/test/unit/models/user_spec.js:
```diff
@@ -0,0 +1,45 @@
+    models = require('../../../server/models'),
+    sandbox = sinon.sandbox.create();
+
+describe('Unit: models/user', function () {
```
🎉

ErisDS commented on core/server/models/client.js:
```diff
@@ -16,9 +16,7 @@ Client = ghostBookshelf.Model.extend({
             uuid: uuid.v4(),
-            secret: secret,
-            status: 'development',
-            type: 'ua'
```
I'm not entirely sure about these changes.

Bookshelf has a concept of saved & unsaved models. The reason for including the defaults in the model, is so that an unsaved model has the default values before being sent to the DB - so that the values can be validated and are present for any logic that happens.

I do not know for sure if we have cases like this, but I think this reasoning is worth considering?

kirrg001 replied:
Thanks for the hint with unsaved models. Will double check 👍

kirrg001 replied:
So `defaults` are called during the `save` call in bookshelf.
[See](https://github.com/bookshelf/bookshelf/blob/0.10.4/src/model.js#L941) bookshelf code section.

They are then available in the event hooks, **not earlier**. Validation doesn't have to happen for defaults - these are (mostly static) values, which we define. But if there is any connected logic it would be helpful to have access to them when *inserting* resources. Right now there is no connected logic, but i will still revert and leave a clear comment. We have to live with the fact that we have to define defaults twice (once in the schema, once in the model layer) - for now.

ErisDS commented on core/server/models/user.js:
```diff
@@ -38,11 +38,9 @@ User = ghostBookshelf.Model.extend({
-        return _.merge({
+        return {
             password: security.identifier.uid(50)
-        }, baseDefaults);
```
All of this weird logic to include the uuid is definitely not needed anymore though! Good spot 👀

ErisDS commented on core/server/models/user.js:
```diff
@@ -201,18 +201,13 @@ User = ghostBookshelf.Model.extend({
     // This is used to bypass validation during the credential check, and must never be done with user-provided data
     // Should be removed when #3691 is done
```
Is this comment still relevant? 😁

kirrg001 replied:
Hmm i think so. It's talking about passing `options.validate=false`.
🤔 Not sure where it's used though - will double check.

kirrg001 replied:
This was originally added when login in, but the e.g. website field was invalid, because we added proper validation after adding the field.

We can remove the option to bypass validation. The validation layer should only validate fields which change.
And if we add a new model, it should check all fields. Commit is coming.

ErisDS reviewed (commented):
I have one thing to consider here - everything else looks 👍 👍

kirrg001 committed:
Differentiate between update and insert when validating model against schema

- update
  - only check `changed` fields
- add
  - check all fields and catch not nullable fields

kirrg001 committed:
revert removing defaults on model layer

kirrg001 committed:
fix tests

kirrg001 committed:
revert reorder

kirrg001 committed:
name functions

kirrg001 committed:
Removed extra safe catch when logging in

- no need to catch the validation error anymore
- on login: we only validate the fields which have changed (status, last_seen)

kirrg001 committed:
fix lint

kirrg001 committed:
Reorder mysql support for knex mock

kirrg001 committed:
fix lint
